### PR TITLE
Transactors

### PR DESCRIPTION
**Pull request by jameshaydon**
Transactor re-definition replaces eval re-difinition.

This is a much simpler model. At first `tx` is set to the identity function, that is, the machine accepts all input expressions as is, and evaluates them.

One can then redefine `tx`, to restrict the inputs to the machine. If `tx` throws an exception, then the input is not accepted, otherwise it can perform effects itself (e.g. mutate some refs), and then return an expression for evaluation on the machine.

See https://github.com/radicle-dev/transparent-machines-doc for more information.

### CODE REVIEW

File: src/Radicle/Internal/Eval.hs [hunk 1]
```diff
@@ -37,21 +37,18 @@ baseEval val = logValPos val $ case val of
...
-            List [val', newSt] -> do
-                setBindings newSt
-                pure val'
-            _ -> throwErrorHere
-               $ OtherError "eval: should return list with value and new env"
-
-
+-- |
+-- A transaction happens as follows:
+-- - First @tx@ is resovled, this is expected to be invocable.
```
Comment (jkarni) [inline src/Radicle/Internal/Eval.hs]:
```suggestion
-- - First @tx@ is resolved, this is expected to be invocable.
```

Comment (jkarni) [inline src/Radicle/Internal/Eval.hs]:
```suggestion
-- - First @tx@ is resolved, this is expected to be invocable.
```

File: test/spec/Radicle/Tests.hs [hunk 2]
```diff
@@ -155,7 +155,7 @@ test_eval =
         f "(insert (ref 0) 1 {})"
         f "{(ref 0) 1}"
         "(lookup :k {'(fn [x] y) 1 :k :v})" `succeedsWith` Keyword [ident|v|]
-        f "(eval {'(fn [y] y) :a-fun} (get-current-state))"
+        --f "(tx {'(fn [y] y) :a-fun} (get-current-state))"
```
Comment (jkarni) [inline test/spec/Radicle/Tests.hs]: huh?

**Comment by jameshaydon** [comment]
Tutorial is now checked in CI:

https://console.cloud.google.com/cloud-build/builds/014715c9-4037-47c2-9de3-39768fa0bf78?project=opensourcecoin

File: rad/prelude/test-eval.rad [hunk 3]
```diff
@@ -1,33 +1,28 @@
...
 (def tests
   "Reference that collects tests defined with the `:test` macro. This
   value is updated by a custom `eval`."
   (ref []))
 
 ;; Annoyingly verbose since we have neither the prelude nor pattern-matching.
-(def eval
-  "Evaluation function that adds :test macro to register tests."
+(def tx
+  "Transaction function that adds :test macro to register tests."
```
Comment (MeBrei) [inline rad/prelude/test-eval.rad]: In primFns it is called `transactor function` Maybe change one of them for consistency.

Comment (jameshaydon) [reply rad/prelude/test-eval.rad]: Actually this changes to a macro in the macros PR so I will just leave this.

**Pull request merged**